controlplane/kubeadm/controllers/kubeadm_control_plane_controller.go

```diff
@@ -347,41 +324,117 @@ func (r *KubeadmControlPlaneReconciler) updateStatus(ctx context.Context, kcp *c
 	return nil
 }
 
-func (r *KubeadmControlPlaneReconciler) upgradeControlPlane(ctx context.Context, cluster *clusterv1.Cluster, kcp *controlplanev1.KubeadmControlPlane, requireUpgrade []*clusterv1.Machine) error {
+func (r *KubeadmControlPlaneReconciler) upgradeControlPlane(ctx context.Context, cluster *clusterv1.Cluster, kcp *controlplanev1.KubeadmControlPlane, logger logr.Logger) (ctrl.Result, error) {
+	// TODO: handle reconciliation of etcd members and kubeadm config in case they get out of sync with cluster
+	ownedMachines, err := r.managementCluster.GetMachinesForCluster(ctx, clusterKey(cluster), internal.OwnedControlPlaneMachines(kcp.Name))
+	if err != nil {
+		logger.Error(err, "failed to retrieve machines for cluster")
+		return ctrl.Result{}, err
+	}
+	requireUpgrade := internal.UnionFilterMachines(
+		ownedMachines,
+		internal.HasOutdatedConfiguration(hash.Compute(&kcp.Spec)),
+		internal.OlderThan(kcp.Spec.UpgradeAfter),
+	)
+	if len(requireUpgrade) == 0 {
+		logger.Info("no machines found to upgrade")
+		// requeue to perform any additional operations that may be required
+		return ctrl.Result{Requeue: true}, nil
+	}
 
-	// TODO: verify health for each existing replica
-	// TODO: mark an old Machine via the label kubeadm.controlplane.cluster.x-k8s.io/selected-for-upgrade
-	// TODO: check full cluster health
-	// TODO: provision new Machine to replace marked Machine
-	// TODO: wait for health of all existing replicas, ideally in a re-entrant way to avoid waiting in process
-	// TODO: wait for full cluster health, ideally in a re-entrant way to avoid waiting in process
-	// TODO: Remove etcd membership for old Machine instance
-	// TODO: Update the kubeadm configmap
-	// TODO: Delete the Marked ControlPlane machine
-	// TODO: Continue with next OldMachine
+	if err := r.healthCheck(ctx, clusterKey(cluster), kcp.Name); err != nil {
+		logger.Error(err, "waiting for control plane to pass health check before continuing upgrade")
+		r.recorder.Eventf(kcp, corev1.EventTypeWarning, "ControlPlaneUnhealthy", "Waiting for control plane to pass health check before continuing upgrade: %v", err)
+		return ctrl.Result{RequeueAfter: HealthCheckFailedRequeueAfter}, err
+	}
 
-	return nil
+	// If there is not already a Machine that is marked for upgrade, find one and mark it
+	selectedForUpgrade := internal.FilterMachines(requireUpgrade, internal.SelectedForUpgrade())
+	if len(selectedForUpgrade) == 0 {
+		selectedMachine, err := r.selectMachineForUpgrade(ctx, cluster, requireUpgrade)
+		if err != nil {
+			logger.Error(err, "failed to select machine for upgrade")
+			return ctrl.Result{}, err
+		}
+		selectedForUpgrade = append(selectedForUpgrade, selectedMachine)
+	}
+
+	// Determine if we need to add an additional control plane Machine or clean up an already replaced one
+	switch {
```

Comment:
this looks like it would be cleaner without a switch

The early return is the first case
then do everything in the default case without the need for additional indentation

it's never too late to add the switch later if it turns out there are more cases here